### ISSUE TITLE
[timoberger repo] Local calendar creation problem

### ISSUE THREAD

**Issue by opendev0** [gggard/AndroidCaldavSyncAdapater#96]
As I tried to find out what's my problem I found DavCalendar throwing an Exception in line 274 of the current master branch, if DavCalendar.createNewAndroidCalendar returns no DavCalendar. It is indeed catched in the SyncAdapter but I think this is not intended?

Unfortunately I didn't exactly find out what my problem is, but it seems to have to do with calendar creation in Android. Can you explain me what you meant by the following comment in line 346-347 of DavCalendar.java?

``` java
// it is possible that this calendar already exists but the provider failed to find it within isCalendarExist()
// the adapter would try to create a new calendar but the provider fails again to create a new calendar.
```

Or maybe you know what someone can do, if ContentProviderClient.insert() returns null (the block before)? That's exactly the reason why I can't sync my calendars...

Thanks in advance.

**Comment by timoberger** [comment]
i made some major redesign within the last commits and it seems that i forgot this point. i am going to fix this soon.

the source of this problem is currently unknown. if you kill the current task (with any task manager), the error will simply disapear.

**Comment by opendev0** [comment]
I saw that and really appreciate it. Was just a little thing I noticed.

With an older version of Android I was able to sync my calendars, just found it odd. I will see if i can find out something in the coming days.

**Comment by timoberger** [comment]
@opendev0 can you test if the last commit fits for you?

**Comment by opendev0** [comment]
No, doesn't help. My androidCalList is empty and I get the exact same error.

**Comment by timoberger** [comment]
can you send me the debug output please?

**Comment by opendev0** [comment]
Here it is (some values were replaced by {VALUE}):

```
09-10 21:49:50.102  23854-23879/org.gege.caldavsyncadapter V/SyncAdapter﹕ onPerformSync() on {USERNAME} with URL https://{MYSERVERIP}/owncloud/remote.php/caldav
09-10 21:49:50.152  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthState: nothing to do
09-10 21:49:50.152  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthScope: nothing to do
09-10 21:49:50.392  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthState: new with user {USERNAME}
09-10 21:49:50.392  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthScope: new
09-10 21:49:50.822  23854-23879/org.gege.caldavsyncadapter I/CaldavFacade﹕ 0 calendars found in set https://{MYSERVERIP}/owncloud/remote.php/caldav
09-10 21:49:50.852  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthState: restored with user {USERNAME}
09-10 21:49:50.852  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthScope: restored
09-10 21:49:51.172  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ Found userPrincipal: https://{MYSERVERIP}/owncloud/remote.php/caldav/principals/{USERNAME}/
09-10 21:49:51.192  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthState: restored with user {USERNAME}
09-10 21:49:51.192  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthScope: restored
09-10 21:49:51.642  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ 1 calendar-home-set found in https://{MYSERVERIP}/owncloud/remote.php/caldav/principals/{USERNAME}/
09-10 21:49:51.692  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthState: restored with user {USERNAME}
09-10 21:49:51.692  23854-23879/org.gege.caldavsyncadapter D/CaldavFacade﹕ LastAuthScope: restored
09-10 21:49:52.172  23854-23879/org.gege.caldavsyncadapter I/CaldavFacade﹕ 2 calendars found in set https://{MYSERVERIP}/owncloud/remote.php/caldav/calendars/{USERNAME}/
09-10 21:49:52.172  23854-23879/org.gege.caldavsyncadapter I/SyncAdapter﹕ 0 calendars found at android
09-10 21:49:52.172  23854-23879/org.gege.caldavsyncadapter I/SyncAdapter﹕ Detected calendar name=Calendar1 URI=https://{MYSERVERIP}/owncloud/remote.php/caldav/calendars/{USERNAME}/private/
09-10 21:49:52.192  23854-23879/org.gege.caldavsyncadapter E/SyncAdapter﹕ failed to get an existing or create a new calendar
09-10 21:49:52.242  23854-23879/org.gege.caldavsyncadapter I/SyncAdapter﹕ Detected calendar name=Calendar2 URI=https://{MYSERVERIP}/owncloud/remote.php/caldav/calendars/{USERNAME}/school/
09-10 21:49:52.252  23854-23879/org.gege.caldavsyncadapter E/SyncAdapter﹕ failed to get an existing or create a new calendar
09-10 21:49:52.292      462-562/system_process D/SyncManager﹕ failed sync operation {USERNAME} u0 (org.gege.caldavsyncadapter.account), com.android.calendar, SERVER, earliestRunTime 17982032, reason: AutoSync, SyncResult: stats [ numIoExceptions: 2]
```

**Comment by opendev0** [comment]
I didn't want to close it -.-

**Comment by timoberger** [comment]
this is a strange behavior.
do the 2 calendars (calendar1/privat and calendar2/school) already exist on your device (are you able to see them within a calendar app)?

**Comment by timoberger** [comment]
if you are able to debug this issue, that would be great
or maybe you could simply remove a) the caldav account and b) the sync adapter itself and then c) reinstall and d) reconfigure everything
or did you try to restart your device?

**Comment by opendev0** [comment]
Already tried all of this and I don't see them in my calendar app. I reinstalled your alpha-2 and the latest master available in F-Droid. They both fail at the same line because ContentClientProvider.insert() returns null.
I also checked the generated URI and contentValues that a passed to this method but they look fine to me.
Maybe I have some kind of permission issue on my device, since Cyanogenmod has lately added SeLinux. I will see if I can find something out.

**Comment by opendev0** [comment]
Problem solved!
After reading about providers and looking in my calendar-provider's debug outputs I realized that it had to be somehow related to permissions:

```
09-11 12:28:47.979    2091-5222/com.android.providers.calendar D/CalendarProvider2﹕ Calendar query from application using privacy guard! pid=6810
```

It turned out to be the integrated incognito-function of CyanogenMod. I set it to be enabled for newly installed apps, so it also prevented CaldavSyncAdapter from transmitting my information to the provider...

I think this is one of the few reasons that provider.insert() actually returns null.
